pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,300 @@
+	if r.Infra.Status.PlatformStatus == nil {
+		return errPlatformStatusNil
+	}
+
+	// Checking whether InfraCluster object exists. If it doesn't, create it.
+	if err := r.Get(ctx, client.ObjectKeyFromObject(target), target); err != nil && !cerrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get InfraCluster: %w", err)
+	} else if err == nil {
+		// When the object already exists, there's nothing to do.
+		return nil
+	}
+
+	if err := r.createAzureInfraCluster(ctx, target, log); err != nil {
+		return fmt.Errorf("failed to create Azure Infra Cluster: %w", err)
+	}
+
+	return nil
+}
+
+// createNewAzureInfraCluster creates a new Azure Infra Cluster.
+//
+//nolint:funlen
+func (r *InfraClusterController) createAzureInfraCluster(ctx context.Context, target *azurev1.AzureCluster, log logr.Logger) error {
+	apiURL, err := url.Parse(r.Infra.Status.APIServerInternalURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse apiUrl: %w", err)
+	}
+
+	port, err := strconv.ParseInt(apiURL.Port(), 10, 32)
+	if err != nil {
+		return fmt.Errorf("failed to parse apiUrl port: %w", err)
+	}
+
+	providerSpec, err := getAzureMAPIProviderSpec(ctx, r.Client)
+	if err != nil {
+		return fmt.Errorf("error obtaining Azure Provider Spec: %w", err)
+	}
+
+	location, err := r.getAzureLocation(ctx)
+	if err != nil {
+		return fmt.Errorf("error obtaining Azure Cluster location: %w", err)
+	}
+
+	log.Info(fmt.Sprintf("AzureCluster %s/%s does not exist, creating it", target.Namespace, target.Name))
+
+	target = &azurev1.AzureCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.Infra.Status.InfrastructureName,
+			Namespace: defaultCAPINamespace,
+			// The ManagedBy Annotation is set so CAPI infra providers ignore the InfraCluster object,
+			// as that's managed externally, in this case by this controller.
+			Annotations: map[string]string{
+				clusterv1.ManagedByAnnotation: managedByAnnotationValueClusterCAPIOperatorInfraClusterController,
+			},
+		},
+
+		Spec: azurev1.AzureClusterSpec{
+			AzureClusterClassSpec: azurev1.AzureClusterClassSpec{
+				Location:         location,
+				AzureEnvironment: "AzurePublicCloud",
+				IdentityRef: &corev1.ObjectReference{
+					Name:      r.Infra.Status.InfrastructureName,
+					Namespace: defaultCAPINamespace,
+					Kind:      "AzureClusterIdentity",
+				},
+			},
+			NetworkSpec: azurev1.NetworkSpec{
+				NodeOutboundLB: &azurev1.LoadBalancerSpec{
+					Name: r.Infra.Status.InfrastructureName,
+					BackendPool: azurev1.BackendPool{
+						Name: r.Infra.Status.InfrastructureName,
+					},
+				},
+				Vnet: azurev1.VnetSpec{
+					Name:          providerSpec.Vnet,
+					ResourceGroup: providerSpec.NetworkResourceGroup,
+				},
+			},
+			ResourceGroup: providerSpec.ResourceGroup,
+			ControlPlaneEndpoint: clusterv1.APIEndpoint{
+				Host: apiURL.Hostname(),
+				Port: int32(port),
+			},
+		},
+	}
```

Comment:
This could be its own function just to simplify things here
```
func (r *InfraClusterController) newAzureCluster(providersSpec machinev1.AzureProviderConfig, apiUrl, port, location string) *azurev1.AzureCluster {
	return &azurev1.AzureCluster{
		ObjectMeta: metav1.ObjectMeta{
			Name:      r.Infra.Status.InfrastructureName,
			Namespace: defaultCAPINamespace,
			// The ManagedBy Annotation is set so CAPI infra providers ignore the InfraCluster object,
			// as that's managed externally, in this case by this controller.
			Annotations: map[string]string{
				clusterv1.ManagedByAnnotation: managedByAnnotationValueClusterCAPIOperatorInfraClusterController,
			},
		},

		Spec: azurev1.AzureClusterSpec{
			AzureClusterClassSpec: azurev1.AzureClusterClassSpec{
				Location:         location,
				AzureEnvironment: "AzurePublicCloud",
				IdentityRef: &corev1.ObjectReference{
					Name:      r.Infra.Status.InfrastructureName,
					Namespace: defaultCAPINamespace,
					Kind:      "AzureClusterIdentity",
				},
			},
			NetworkSpec: azurev1.NetworkSpec{
				NodeOutboundLB: &azurev1.LoadBalancerSpec{
					Name: r.Infra.Status.InfrastructureName,
					BackendPool: azurev1.BackendPool{
						Name: r.Infra.Status.InfrastructureName,
					},
				},
				Vnet: azurev1.VnetSpec{
					Name:          providerSpec.Vnet,
					ResourceGroup: providerSpec.NetworkResourceGroup,
				},
			},
			ResourceGroup: providerSpec.ResourceGroup,
			ControlPlaneEndpoint: clusterv1.APIEndpoint{
				Host: apiURL.Hostname(),
				Port: int32(port),
			},
		},
	}
}
```